coc/abc.py

```diff
@@ -201,42 +208,72 @@ def _load_json_meta(cls, troop_meta, id, name, lab_to_townhall):
             cls.is_elixir_spell = True
         elif production_building == "Mini Spell Factory":
             cls.is_dark_spell = True
+        elif name in coc.HERO_PETS_ORDER:
+            production_building = "Pet Shop"
 
         # load buildings
         with open(BUILDING_FILE_PATH) as fp:
             buildings = ujson.load(fp)
 
-        # without production_building, it is a hero or pet
+        # without production_building, it is a hero
         if not production_building:
             laboratory_levels = troop_meta.get("LaboratoryLevel")
         else:
-            # it is a troop or spell
+            # it is a troop or spell or siege
             prod_unit = buildings.get(production_building)
-            min_prod_unit_level = troop_meta.get("BarrackLevel", [None, ])[0]
-            # there are some special troops, which have no BarrackLevel attribute
-            if not min_prod_unit_level:
-                laboratory_levels = troop_meta.get("LaboratoryLevel")
-            else:
+            if "barrack" in production_building.lower() or "Spell Forge" == production_building or \
+                    "SiegeWorkshop" == production_building or "Mini Spell Factory" == production_building:
+                min_prod_unit_level = troop_meta.get("BarrackLevel", [None, ])[0]
+                # there are some special troops, which have no BarrackLevel attribute
+                if not min_prod_unit_level:
+                    laboratory_levels = troop_meta.get("LaboratoryLevel")
+                else:
+                    # get the min th level were we can unlock by the required level of the production building
+                    min_th_level = [th for i, th in
```

Comment:
Slick enumeration, but I see it used further down. Let's consider making this its own function too to adhere to the DRY principle.